src/cloud-sql-instance.ts

```diff
@@ -67,14 +72,54 @@ export class CloudSQLInstance {
     authType,
     instanceConnectionName,
     sqlAdminFetcher,
+    limitRateInterval = 30 * 1000, // 30s default
   }: CloudSQLInstanceOptions) {
-    this.ipType = ipType;
     this.authType = authType;
     this.instanceInfo = parseInstanceConnectionName(instanceConnectionName);
+    this.ipType = ipType;
+    this.limitRateInterval = limitRateInterval;
     this.sqlAdminFetcher = sqlAdminFetcher;
   }
 
+  // p-throttle library has to be initialized in an async scope in order to
+  // use dynamic import so that it's also compatible with CommonJS
+  private async initializeRateLimiter() {
+    if (this.throttle) {
+      return;
+    }
+    const pThrottle = (await import('p-throttle')).default;
+    this.throttle = pThrottle({
+      limit: 1,
+      interval: this.limitRateInterval,
+      strict: true,
+    }) as ReturnType<typeof pThrottle>;
+  }
+
+  async forceRefresh(): Promise<void> {
+    // if a refresh is already ongoing, just await for its promise to fulfill
+    // so that a new instance info is available before reconnecting
+    if (this.ongoingRefreshPromise) {
+      await this.ongoingRefreshPromise;
+      return;
+    }
+    this.cancelRefresh();
+    return this.refresh();
+  }
+
   async refresh(): Promise<void> {
+    this.ongoingRefreshPromise = this.throttle
+      ? this.throttle(this._refresh).call(this)
+      : this._refresh();
+    await this.ongoingRefreshPromise;
+    this.ongoingRefreshPromise = undefined;
```

Comment:
let's add a comment or two for this block of code as it is important part of the logic